dll/win32/shell32/dialogs/general.cpp

```diff
@@ -91,32 +84,60 @@ SHELL32_ReadRegShellState(PREGSHELLSTATE prss)
     return err == ERROR_SUCCESS && prss->dwSize >= REGSHELLSTATE_SIZE;
 }
 
-// bIconUnderline is TRUE if "Underline icon titles only when I point at them".
-// bIconUnderline is FALSE if "Underline icon titles consistent with my browser".
+// bUnderlineHover is TRUE if "Underline icon titles only when I point at them".
+// bUnderlineHover is FALSE if "Underline icon titles consistent with my browser".
 //////////////////////////////////////////////////////////////////////////////
 // API Monitor:
-// SHELL32.dll	SHRegGetUSValueW ( "Software\Microsoft\Windows\CurrentVersion\Explorer", "IconUnderline", NULL, 0x00d2f324, 0x00d2f328, FALSE, 0x00d2f32c, 4 )	ERROR_SUCCESS		0.0002484
-// SHELL32.dll	IsDlgButtonChecked ( 0x0005009e, 30104 )	BST_CHECKED		0.0000212
-// SHELL32.dll	SHRegSetUSValueW ( "Software\Microsoft\Windows\CurrentVersion\Explorer", "IconUnderline", 0, 0x00d2f314, 4, 6 )	ERROR_SUCCESS		0.0008300
-// Explorer.EXE	SHSettingsChanged ( 0, "Software\Microsoft\Windows\CurrentVersion\Explorer\IconUnderline" )			0.0000092
-static BOOL IntSetUnderlineState(BOOL bIconUnderline)
+// SHELL32.dll	SHRegGetUSValueW ( "Software\Microsoft\Windows\CurrentVersion\Explorer", "IconUnderline", NULL, 0x00d2f324, 0x00d2f328, FALSE, 0x00d2f32c, 4 )	ERROR_SUCCESS
+// SHELL32.dll	IsDlgButtonChecked ( 0x0005009e, 30104 )	BST_CHECKED
+// SHELL32.dll	SHRegSetUSValueW ( "Software\Microsoft\Windows\CurrentVersion\Explorer", "IconUnderline", 0, 0x00d2f314, 4, 6 )	ERROR_SUCCESS
+// Explorer.EXE	SHSettingsChanged ( 0, "Software\Microsoft\Windows\CurrentVersion\Explorer\IconUnderline" ) (WM_SETTINGCHANGE handling)
+static BOOL IntSetUnderlineState(BOOL bUnderlineHover)
 {
     LSTATUS Status;
-    DWORD dwValue = (bIconUnderline ? 3 : 2), dwSize = sizeof(DWORD);
+    DWORD dwValue = bUnderlineHover ? UNDERLINE_HOVER : UNDERLINE_IE, dwSize = sizeof(DWORD);
     Status = SHRegSetUSValue(s_pszExplorerKey, L"IconUnderline", REG_NONE,
                              &dwValue, dwSize, SHREGSET_FORCE_HKCU | SHREGSET_HKLM);
     if (Status != ERROR_SUCCESS)
         return FALSE;
 
-    SHSendMessageBroadcastW(WM_SETTINGCHANGE, 0, (LPARAM)L"Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\IconUnderline");
+    SHSendMessageBroadcastW(WM_SETTINGCHANGE, 0, (LPARAM)(REGSTR_PATH_EXPLORER L"\\IconUnderline"));
     return TRUE;
 }
 
-static BOOL IntGetUnderlineState(VOID)
+static UINT IntGetRawIconUnderlineValue(VOID)
 {
-    DWORD dwValue, dwDefault = 2, dwSize = sizeof(DWORD);
+    DWORD dwValue, dwDefault = UNDERLINE_ON, dwSize = sizeof(DWORD);
     SHRegGetUSValue(s_pszExplorerKey, L"IconUnderline", NULL, &dwValue, &dwSize, FALSE, &dwDefault, sizeof(DWORD));
-    return dwValue == 3;
+    return dwValue;
+}
+
+static UINT SHELL_GetIconUnderlineMode()
+{
+    UINT mode = IntGetRawIconUnderlineValue();
+    if (mode < UNDERLINE_IE)
+        return mode;
+
+    WCHAR buf[42];
```

Comment:
Since you care about "no" or "hover" only, you could use a reduced size: 7 characters (== 5 + 1 + terminating NUL, where "5" is the length of "hover"; the extra "1" is to detect if there is something more following the longest case, e.g. if you have something like "hoverX" it would have to be excluded and the UNDERLINE_ON default be taken instead).